login/service.go

```diff
@@ -354,14 +368,32 @@ func (keycloak *KeycloakOAuthProvider) linkProvider(ctx linkInterface, req *goa.
 	return ctx.TemporaryRedirect()
 }
 
-func (keycloak *KeycloakOAuthProvider) saveReferrer(ctx context.Context, state uuid.UUID, referrer string) error {
+func (keycloak *KeycloakOAuthProvider) saveReferrer(ctx linkInterface, state uuid.UUID, referrer string, validReferrerURL string) error {
+	matched, err := regexp.MatchString(validReferrerURL, referrer)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"referrer":           referrer,
+			"valid-referrer-url": validReferrerURL,
+			"err":                err,
+		}, "Can't match referrer and whitelist regex")
+		jerrors, _ := jsonapi.ErrorToJSONAPIErrors(goa.ErrInternal(err.Error()))
+		return ctx.InternalServerError(jerrors)
+	}
+	if !matched {
+		log.Error(ctx, map[string]interface{}{
+			"referrer":           referrer,
+			"valid-referrer-url": validReferrerURL,
+		}, "Referrer not valid")
+		jerrors, _ := jsonapi.ErrorToJSONAPIErrors(goa.ErrBadRequest("Not valid redirect URL"))
+		return ctx.BadRequest(jerrors)
+	}
 	// TODO The state reference table will be collecting dead states left from some failed login attempts.
 	// We need to clean up the old states from time to time.
 	ref := auth.OauthStateReference{
 		ID:       state,
 		Referrer: referrer,
 	}
-	err := application.Transactional(keycloak.db, func(appl application.Application) error {
+	err = application.Transactional(keycloak.db, func(appl application.Application) error {
 		_, err := appl.OauthStates().Create(ctx, &ref)
 		return err
```

Comment:
You mean below? Yes that's true, but it's just matter of you taste, right? :) It doesn't change anything.